// @carbon/react/lib/components/Toggle/Toggle.Skeleton.js
/**
 * Copyright IBM Corp. 2016, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _rollupPluginBabelHelpers = require('../../_virtual/_rollupPluginBabelHelpers.js');
var PropTypes = require('prop-types');
var React = require('react');
var cx = require('classnames');
var usePrefix = require('../../internal/usePrefix.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var cx__default = /*#__PURE__*/_interopDefaultLegacy(cx);

class ToggleSkeleton extends React__default["default"].Component {
  render() {
    var _div;

    const {
      id,
      labelText,
      className,
      size,
      ...rest
    } = this.props;
    return /*#__PURE__*/React__default["default"].createElement(usePrefix.PrefixContext.Consumer, null, prefix => {
      const toggleInputClassNames = cx__default["default"](`${prefix}--toggle ${prefix}--skeleton`, {
        [`${prefix}--toggle-input--small`]: size === 'sm'
      });
      return /*#__PURE__*/React__default["default"].createElement("div", _rollupPluginBabelHelpers["extends"]({
        className: cx__default["default"](`${prefix}--form-item`, className)
      }, rest), /*#__PURE__*/React__default["default"].createElement("input", {
        type: "checkbox",
        id: id,
        className: toggleInputClassNames
      }), /*#__PURE__*/React__default["default"].createElement("label", {
        className: `${prefix}--toggle-input__label`,
        htmlFor: id,
        "aria-label": labelText ? undefined : this.props['aria-label']
      }, labelText ? _div || (_div = /*#__PURE__*/React__default["default"].createElement("div", null, labelText)) : null, /*#__PURE__*/React__default["default"].createElement("span", {
        className: `${prefix}--toggle__switch`
      }, /*#__PURE__*/React__default["default"].createElement("span", {
        className: `${prefix}--toggle__text--left`
      }), /*#__PURE__*/React__default["default"].createElement("span", {
        className: `${prefix}--toggle__appearance`
      }), /*#__PURE__*/React__default["default"].createElement("span", {
        className: `${prefix}--toggle__text--right`
      }))));
    });
  }

}

_rollupPluginBabelHelpers.defineProperty(ToggleSkeleton, "propTypes", {
  ['aria-label']: PropTypes__default["default"].string.isRequired,

  /**
   * Specify an optional className to add to the form item wrapper.
   */
  className: PropTypes__default["default"].string,

  /**
   * Provide an id that unique represents the underlying `<input>`
   */
  id: PropTypes__default["default"].string,

  /**
   * Provide the text that will be read by a screen reader when visiting this
   * control
   * `aria-label` is always required but will be undefined if `labelText` is also
   * provided
   */
  labelText: PropTypes__default["default"].string,

  /**
   * Specify the size of the Toggle. Currently only supports 'sm' or 'md' (default)
   */
  size: PropTypes__default["default"].oneOf(['sm', 'md'])
});

_rollupPluginBabelHelpers.defineProperty(ToggleSkeleton, "defaultProps", {
  ['aria-label']: 'Toggle is loading',
  size: 'md'
});

exports.ToggleSkeleton = ToggleSkeleton;
exports["default"] = ToggleSkeleton;
